### PR TITLE
Add auto metric support for GA4 & BigQuery

### PR DESCRIPTION
**Pull request by mknowlton89**
### Features and Changes

This PR adds auto metric support for Google Analytics 4. In order to support this, there were a few changes that were necessary within `SQLIntegration` due to the inherent differences between how Google Analytics 4 and the existing event trackers with auto metric support (Segment & Rudderstack) store tracked events.

Segment & Rudderstack have a single table that outlines all of the unique events that are tracked, and then each row of that tables gets it's own table where each instance of that event are recorded.

Google Analytics 4, however, stores all tracked events on a single table (kinda, they're stored on intraday tables, but all share the same schema).

In order to make auto metrics work with both approaches, we had to introduce an `eventHasUniqueTable` param within `SchemaFormatConfig` and we also had to introduce a `groupByColumns` param within the same config.

These two params ultimately power how we're able to support both approaches as the underlying queries are materially different, especially the actual metric queries. With Segment/Rudderstack, you could query the event's table directly (e.g. if the tracked event was "page_viewed" the FROM clause could be `FROM datasource.schema.page_viewed`), but with Google Analytics 4, all events live on the same table, so to build the metric, you need a `WHERE` clause. (e.g. `WHERE event_name in ('page_viewed')`).

### Testing
- [x] Create a new GA4 data source and ensure you can create auto metrics correctly
- [x] Create a new Segment data source and ensure you can create auto metrics correctly without any regressions

### Other Notes
- If this [PR](https://github.com/growthbook/growthbook/pull/1419) goes out first, I will update the docs in this PR to update the list of event trackers that support auto metrics.
- Additionally, we've made the decision that moving forward, creating auto metrics will be done via the Metrics page, or the datasource/id page via the "Discover Metrics" button. As such, this PR doesn't include any updates to the `NewDataSourceForm` to allow orgs to opt-in to auto metrics when adding a new GA4 data source. If this [PR](https://github.com/growthbook/growthbook/pull/1419) has not been merged and you'd like to test this PR locally, you can do so by changing `const hasStep2 = true` within `NewDataSourceForm.tsx`. Otherwise, if the PR linked has been merged, you can create a GA4 data source and then use the "Discover Metrics" button. If you need a sample GA4 data source to test this with, please just reach out to me (@mknowlton89) via Slack.

### CODE REVIEW

File: packages/back-end/src/integrations/SqlIntegration.ts [hunk 1]
```diff
@@ -1542,12 +1566,25 @@ export default abstract class SqlIntegration
...
+      groupByColumns,
+      eventHasUniqueTable,
     } = this.getSchemaFormatConfig(schemaFormat);
 
     const currentDateTime = new Date();
-    const SevenDaysAgo = new Date(
+    const sevenDaysAgo = new Date(
       currentDateTime.valueOf() - 7 * 60 * 60 * 24 * 1000
     );
+    const startYear = sevenDaysAgo.getFullYear();
```
Comment (mknowlton89) [inline packages/back-end/src/integrations/SqlIntegration.ts:1577]: This is very verbose - open to suggestions

Comment (bttf) [reply packages/back-end/src/integrations/SqlIntegration.ts:1577]: We have the `date-fns` package available on the back-end. It should help accomplishing this easier

- https://date-fns.org/v2.30.0/docs/sub 
- https://date-fns.org/v2.30.0/docs/format

Comment (mknowlton89) [reply packages/back-end/src/integrations/SqlIntegration.ts]: @bttf Great suggestion - thank you! I've pushed up changes utilizing `date-fns` - way less verbose and way better readability. I also added a `dateFormat` property to the `SchemaFormatConfig` for added flexibility down the road.

File: packages/back-end/src/integrations/SqlIntegration.ts [hunk 2]
```diff
@@ -1542,12 +1569,13 @@ export default abstract class SqlIntegration
...
+      eventHasUniqueTable,
+      dateFormat,
     } = this.getSchemaFormatConfig(schemaFormat);
 
-    const currentDateTime = new Date();
-    const SevenDaysAgo = new Date(
-      currentDateTime.valueOf() - 7 * 60 * 60 * 24 * 1000
-    );
+    const today = formatDate(new Date(), dateFormat);
+    const sevenDaysAgo = formatDate(subDays(new Date(), 100), dateFormat);
```
Comment (bttf) [inline packages/back-end/src/integrations/SqlIntegration.ts:1578]: Is `100` right in `subDays(new Date(), 100)`?
```suggestion
    const sevenDaysAgo = formatDate(subDays(new Date(), 100), dateFormat);
```

Comment (mknowlton89) [reply packages/back-end/src/integrations/SqlIntegration.ts]: Oh gosh - no - an artifact from testing. Updated [here](https://github.com/growthbook/growthbook/pull/1423/commits/67f2f5b71a13d26c5507a8eb1448a53f13685438).

**Review by bttf — APPROVED** [review]

**Pull request merged**